actionpack/lib/action_controller/metal/rate_limiting.rb

```diff
@@ -44,19 +49,73 @@ module ClassMethods
       #       RATE_LIMIT_STORE = ActiveSupport::Cache::RedisCacheStore.new(url: ENV["REDIS_URL"])
       #       rate_limit to: 10, within: 3.minutes, store: RATE_LIMIT_STORE
       #     end
-      def rate_limit(to:, within:, by: -> { request.remote_ip }, with: -> { head :too_many_requests }, store: cache_store, **options)
-        before_action -> { rate_limiting(to: to, within: within, by: by, with: with, store: store) }, **options
+      #
+      #     class SessionsController < ApplicationController
+      #       rate_limit to: 10, within: 3.minutes, only: :create, strategy :sliding_window
+      #     end
+      def rate_limit(to:, within:, by: -> { request.remote_ip }, with: -> { head :too_many_requests }, store: cache_store, strategy: :fixed_window, **options)
+        before_action -> { rate_limiting(to: to, within: within, by: by, with: with, store: store, strategy: strategy) }, **options
+      end
+    end
+
+    class BaseStrategy
+      attr_reader :to, :within, :store
+
+      def initialize(to:, within:, store:)
+        @to = to
+        @within = within
+        @store = store
+      end
+
+      def increment_and_check(key:)
+        raise NotImplementedError, "Subclasses must implement `increment_and_check` method"
       end
     end
 
     private
-      def rate_limiting(to:, within:, by:, with:, store:)
-        count = store.increment("rate-limit:#{controller_path}:#{instance_exec(&by)}", 1, expires_in: within)
-        if count && count > to
+      def rate_limiting(to:, within:, by:, with:, store:, strategy:)
+        key = "rate-limit:#{controller_path}:#{instance_exec(&by)}"
+        strategy_class = if strategy.is_a?(Class)
+          strategy
+        elsif STRATEGY_CLASSES.key?(strategy)
+          STRATEGY_CLASSES[strategy]
+        else
+          raise ArgumentError, "Unknown rate limiting strategy: #{strategy}"
+        end
+
+        limiter = strategy_class.new(to: to, within: within, store: store)
+
+        if limiter.increment_and_check(key: key)
           ActiveSupport::Notifications.instrument("rate_limit.action_controller", request: request) do
             instance_exec(&with)
           end
         end
       end
+
+      class FixedWindowStrategy < BaseStrategy
+        def increment_and_check(key:)
+          count = store.increment(key, 1, expires_in: within)
+          count && count > to
+        end
+      end
+
+      class SlidingWindowStrategy < BaseStrategy
+        def increment_and_check(key:)
+          current_time = Time.current.to_f
```

Comment:
I'm wondering if using `Time.current` is the best approach here 🤔

Reply:
`Time.current` _feels_ wrong because if the timezone is defined as anything but UTC there will be jumps during the daylight saving time events but one might argue it’s an over-optimisation. However, it does feel using UTC is not going overboard.

I would favour more though to use a monotonic clock instead and compare the elapsed time in that way.

Reply:
From a quick test in the console it seems you are right!

Reply:
I'll wait if someone has another opinion on the matter.
I might use `Time.now` because is more performant

Reply:
> I would favour more though to use a monotonic clock

That's not going to work when shared datastores (eg Redis) are backing the limiter, since every server your app is running on has their own monotonic clock and these are not comparable with each other.

`Time.now.to_f` is a start but even then time syncing issues can be problematic across different servers. More robust is to use the datastore time functions (eg [TIME](https://redis.io/docs/latest/commands/time/) or [clock_timestamp()](https://www.postgresql.org/docs/current/functions-datetime.html#FUNCTIONS-DATETIME-CURRENT)) instead.

Reply:
> That's not going to work when shared datastores (eg Redis) are backing the limiter, since every server your app is running on has their own monotonic clock and these are not comparable with each other.

Yeah, thought the same thing.

Rails ships with several implementations of cache store, managing time can be tricky for sure.